### PR TITLE
Support multi-tenancy for LocalRegexGuardrail

### LINKED ISSUE #4119: [FEATURE] Add muti-tenancy support to Guardrails
**Is your feature request related to a problem?**
[Guardrail](https://docs.opensearch.org/latest/ml-commons-plugin/remote-models/guardrails/) filtered input / output based on word blacklist or regex pattern match.

Previously, we have introduced multi-tenancy support to connectors, models, agents, etc. However, this paradigm has not been introduced to guardrail. This prohibits its integration into cloud service environments where multiple tenants may share a single instance.

**What solution would you like?**
Introduce multi-tenancy support to guardral

**What alternatives have you considered?**

**Do you have any additional context?**

#3307 made primary setup for multi-tenancy. #3382, #3399, #3416 and #3432 add supports to some other components in ml-commons.

### PR DESCRIPTION
**Pull request by yuancu**
### Description

This PR Introduce add multi-tenancy support to LocalRegexGuardrail by introducing  sdk client to perform search requests.

### Related Issues

Resolves #4119 

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by xinyual** [comment]
LGTM.

**Review by xinyual — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 1]
```diff
@@ -230,32 +235,35 @@ public Boolean validateStopWordsSingleIndex(String input, String indexName, List
...
             } else {
-                client.search(searchRequest, new LatchedActionListener(ActionListener.<SearchResponse>wrap(r -> {
-                    if (r == null || r.getHits() == null || r.getHits().getTotalHits() == null || r.getHits().getTotalHits().value() == 0) {
-                        hitStopWords.set(true);
-                    }
-                }, e -> {
-                    log.error("Failed to search stop words index {}", indexName, e);
-                    hitStopWords.set(true);
-                }), latch));
+                sdkClient
```
Comment (zane-neo) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:239]: It seems the if...else can be merged, and check the other places using `SdkClient`, try to always use the same approach like:
```
 try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
            sdkClient.getDataObjectAsync(getDataObjectRequest).whenComplete((
```
Without this try block, it could cause issue some edge cases.

**Comment by zane-neo** [comment]
Code coverage failed: https://github.com/opensearch-project/ml-commons/pull/4120/checks?check_run_id=48455690124, please take a look and add more tests to cover.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 1 above)
Comment (yuancu) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:239]: By *merging the if...else*, do you mean I can always stash and restore the thread context no matter whether it is a stop word system index or not?

Actually I don't really know why it has to stash and restore the thread context when the index is a system index, but does not have to do so when the index is not. Explanations will be helpful

**Comment by yuancu** [comment]
Hi @zane-neo , I updated implementations and fixed tests. Can you help approve the CI workflows?

**Review by akolarkunnu — APPROVED** [review]
LGTM

**Review by zane-neo — APPROVED** [review]

**Review by zane-neo — APPROVED** [review]

**Review by xinyual — APPROVED** [review]
